### PR TITLE
Remove duplicated code by factoring on endian parameter.

### PR DESCRIPTION
**Pull request by jaykrell**
UTF16 and UTF32 logic is all duplicated just to vary read/write in little/big endian order.
Factor out endianness as a parameter to avoid duplicating the logic.
Force inline to mimic previous performance.
Note that performance can still be improved on most systems by using direct possibly unaligned memory reads/writes instead of byte-wise reads/writes, subject to host ifdef.

### CODE REVIEW

File: mono/eglib/giconv.c [hunk 1]
```diff
@@ -254,8 +254,16 @@ g_iconv (GIConv cd, gchar **inbytes, gsize *inbytesleft,
...
-decode_utf32be (char *inbuf, size_t inleft, gunichar *outchar)
+static FORCE_INLINE (uint32_t)
+read_uint32_endian (unsigned char *inptr, unsigned endian)
+{
+	if (endian == G_LITTLE_ENDIAN)
+		return (inptr[3] << 24) | (inptr[2] << 16) | (inptr[1] << 8) | inptr[0];
+	return (inptr[0] << 24) | (inptr[1] << 16) | (inptr[2] << 8) | inptr[3];
+}
+
+static FORCE_INLINE (int)
```
Comment (marek-safar) [inline mono/eglib/giconv.c]: I don't think it's good idea to force inline this one

Comment (jaykrell) [reply mono/eglib/giconv.c]: Intent was to factor/share source w/o changing object code, but I'm ok either way. The function has to be inlined in order to propagate the constant.

Comment (jaykrell) [reply mono/eglib/giconv.c]: Ah, specifically utf32, which we never even use, as opposed to utf16 which we heavily use?

**Comment by jaykrell** [comment]
@monojenkins build Linux ARMv7 hard float

**Comment by jaykrell** [comment]
@monojenkins build Linux ARMv5 soft float

**Comment by jaykrell** [comment]
@monojenkins build Linux AArch64

**Comment by jaykrell** [comment]
@monojenkins build Linux x64 Acceptance Tests

**Pull request merged**